substrate/frame/election-provider-multi-block/src/signed/tests.rs

```diff
@@ -633,4 +656,648 @@ mod e2e {
+
+				// Clear verifier events to track from this point
+				let _ = verifier_events();
+
+				// Initial verifier state should have no queued solution
+				assert_eq!(VerifierPallet::queued_score(), None);
+
+				// Bad solution is the current leader
+				let mut remaining_leader = Submissions::<Runtime>::leader(current_round).unwrap();
+				assert_eq!(remaining_leader.0, 99);
+
+				// Bad solution starts verification
+				roll_next(); // SignedValidation(2) -> SignedValidation(1)
+				assert!(matches!(VerifierPallet::status(), crate::verifier::Status::Ongoing(_)));
+
+				roll_next(); // SignedValidation(1) -> SignedValidation(0)
+				assert!(matches!(VerifierPallet::status(), crate::verifier::Status::Ongoing(_)));
+
+				roll_next(); // SignedValidation(0) -> Unsigned(4) (verification of bad solution complete,
+				 // verification of the 2nd solution hasn't started yet)
+
+				// Check events after bad solution verification completes
+				let bad_solution_events = verifier_events_since_last_call();
+				assert_eq!(
+					bad_solution_events,
+					vec![
+						crate::verifier::Event::Verified(2, 0),
+						crate::verifier::Event::Verified(1, 0),
+						crate::verifier::Event::Verified(0, 0),
+						crate::verifier::Event::VerificationFailed(
+							0,
+							FeasibilityError::InvalidScore
+						),
+					]
+				);
+
+				// Verify phase transition to unsigned
+				assert!(matches!(MultiBlock::current_phase(), Phase::Unsigned(_)));
+
+				// Check that invalid solution is slashed but good solution remains
+				assert_eq!(Submissions::<Runtime>::submitters_count(current_round), 1);
+				assert!(Submissions::<Runtime>::has_leader(current_round));
+				remaining_leader = Submissions::<Runtime>::leader(current_round).unwrap();
+				assert_eq!(remaining_leader.0, 999); // Good solution remains
+
+				// Check that expected events were emitted for the rejection
+				assert_eq!(
+					signed_events(),
+					vec![
+						Event::Registered(
+							0,
+							99,
+							ElectionScore {
+								minimal_stake: 110,
+								sum_stake: 130,
+								sum_stake_squared: 8650
+							}
+						),
+						Event::Registered(
+							0,
+							999,
+							ElectionScore {
+								minimal_stake: 55,
+								sum_stake: 130,
+								sum_stake_squared: 8650
+							}
+						),
+						Event::Stored(0, 999, 0),
+						Event::Stored(0, 999, 1),
+						Event::Stored(0, 999, 2),
+						Event::Slashed(0, 99, 5),
+					]
+				);
+				// Verify no additional events occurred
+				let remaining_events = verifier_events_since_last_call();
+				assert_eq!(remaining_events, vec![], "No additional verifier events should occur");
+
+				// Verifier should be STOPPED when transitioning to unsigned
+				assert_eq!(VerifierPallet::status(), crate::verifier::Status::Nothing);
+
+				// Verify no solution was queued (verification was stopped, not completed)
+				assert_eq!(VerifierPallet::queued_score(), None);
```

Comment:
this is all good so far. You can also extend it to use `AreWeDoneMode::BackToSigned`, and see that it will go back to
signed phase and then in the second round the good solution is verified.

Reply:
ah this might be the next test :)